### ISSUE TITLE
[EKS] [BUG]: invalid default `tolerations` in CoreDNS addon configuration schema

### ISSUE THREAD

**Issue by iul1an** [aws/containers-roadmap#2167]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Tell us about your request**
The configuration JSON schema for the CoreDNS addon contains an invalid default toleration:
```
➜  aws eks describe-addon-configuration --addon-name coredns --addon-version v1.10.1-eksbuild.4 --query 'configurationSchema' --output text | jq '.definitions.Coredns.properties.tolerations.default[] | select(.key == "node-role.kubernetes.io/master")'
{
  "key": "node-role.kubernetes.io/master",
  "operator": "NoSchedule"
}
```
The `operator` field cannot be `NoSchedule`, it can be either `Exists` or `Equal`:
```
➜  kubectl explain deployment.spec.template.spec.tolerations.operator
GROUP:      apps
KIND:       Deployment
VERSION:    v1

FIELD: operator <string>

DESCRIPTION:
    Operator represents a key's relationship to the value. Valid operators are
    Exists and Equal. Defaults to Equal. Exists is equivalent to wildcard for
    value, so that a pod can tolerate all taints of a particular category.
    
    Possible enum values:
     - `"Equal"`
     - `"Exists"`

```
`NoSchedule` is a valid value for the `effect` field:
```
➜  kubectl explain deployment.spec.template.spec.tolerations.effect  
GROUP:      apps
KIND:       Deployment
VERSION:    v1

FIELD: effect <string>

DESCRIPTION:
    Effect indicates the taint effect to match. Empty means match all taint
    effects. When specified, allowed values are NoSchedule, PreferNoSchedule and
    NoExecute.
    
    Possible enum values:
     - `"NoExecute"` Evict any already-running pods that do not tolerate the
    taint. Currently enforced by NodeController.
     - `"NoSchedule"` Do not allow new pods to schedule onto the node unless
    they tolerate the taint, but allow all pods submitted to Kubelet without
    going through the scheduler to start, and allow all already-running pods to
    continue running. Enforced by the scheduler.
     - `"PreferNoSchedule"` Like TaintEffectNoSchedule, but the scheduler tries
    not to schedule new pods onto the node, rather than prohibiting new pods
    from scheduling onto the node entirely. Enforced by the scheduler.
```

The actual Kubernetes CoreDNS deployment has a valid toleration for the `node-role.kubernetes.io/master` key and it uses `"effect": "NoSchedule"`, so the solution is to replace `"operator": "NoSchedule"` with `"effect": "NoSchedule"` in the configuration schema. Also, the toleration `node-role.kubernetes.io/master:NoSchedule` was [deprecated in favor](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.20.md#no-really-you-must-read-this-before-you-upgrade) of `node-role.kubernetes.io/control-plane:NoSchedule` in Kubernetes 1.20 and [removed](https://kubernetes.io/blog/2022/04/07/upcoming-changes-in-kubernetes-1-24/#api-removals-deprecations-and-other-changes-for-kubernetes-1-24) in v1.24, so maybe the add-on should also reflect that.

**Which service(s) is this request for?**
EKS

**Additional context**
All CoreDNS add-on versions are affected by this bug.
**There is no actual impact, as the actual CoreDNS deployment has valid tolerations, only the schema is faulty.** However, I was trying to add a new toleration on top of the defaults and I was blindly copying the defaults and I kept getting `"Amazon EKS was unable to complete the addon operation. The configuration values provided are invalid."` after submitting the changes. Took me a while to realize the defaults in the schema are invalid and I don't want other people to waste time with this as I did :smile:

**Comment by iul1an** [comment]
The support for setting CoreDNS tolerations was added as a result of this request: https://github.com/aws/containers-roadmap/issues/1930

**Comment by oliviassss** [comment]
@iul1an, thanks for reporting this. This typo actually exists in our doc, which is the source of `aws eks describe-addon-configuration`. If you check the deployment directly, you'll see we have `"effect": "NoSchedule"`
```
% aws eks create-addon --cluster-name test-127 --addon-name coredns --endpoint $BETA --addon-version v1.10.1-eksbuild.4

% k get deployments.apps -n kube-system coredns -o json  | grep -A 3 tolerations
                "tolerations": [
                    {
                        "effect": "NoSchedule",
                        "key": "node-role.kubernetes.io/master"
```
But we will fix the typo in our doc in a coming release.

Since for our addon version v1.9.3-eksbuild.x and v1.10.1-eksbuild.x we support on 1.24+, so we can deprecate the default toleration `node-role.kubernetes.io/master:NoSchedule` and add `node-role.kubernetes.io/control-plane:NoSchedule`. For v1.8.3-eksbuild.x that are on 1.22 - 1.26, we can support both in default toleration for compatibility.

**Comment by oliviassss** [comment]
@iul1an, thanks for reporting this. This typo actually exists in our doc, which is the source of `aws eks describe-addon-configuration`. If you check the deployment directly, you'll see we have `"effect": "NoSchedule"`
```
% aws eks create-addon --cluster-name test-127 --addon-name coredns --endpoint $BETA --addon-version v1.10.1-eksbuild.4

% k get deployments.apps -n kube-system coredns -o json  | grep -A 3 tolerations
                "tolerations": [
                    {
                        "effect": "NoSchedule",
                        "key": "node-role.kubernetes.io/master"
```
But we will fix the typo in our doc in a coming release.

Since for our addon version v1.9.3-eksbuild.x and v1.10.1-eksbuild.x we support on 1.24+, we can deprecate the default toleration `node-role.kubernetes.io/master:NoSchedule` and add `node-role.kubernetes.io/control-plane:NoSchedule`. For v1.8.3-eksbuild.x that are on 1.22 - 1.26, we can support both in default toleration for compatibility.

**Comment by oliviassss** [comment]
@iul1an, the removal of the taint and tolerations on the `node-role.kubernetes.io/master` actually came in 1.25 and 1.26, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-cluster-lifecycle/kubeadm/2067-rename-master-label-taint#renaming-the-node-rolekubernetesiomaster-node-taint

Also in the 1.26 [changelog](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.26.md):

> Kubeadm: removed the toleration for the [node-role.kubernetes.io/master](http://node-role.kubernetes.io/master) taint from the CoreDNS deployment of kubeadm. With the 1.25 release of kubeadm the taint [node-role.kubernetes.io/master](http://node-role.kubernetes.io/master) is no longer applied to control plane nodes and the toleration for it can be removed with the release of 1.26. You can also perform the same toleration removal from your own addon manifests. (#112008, @pacoxu)

So I think we should keep both for coredns v1.9.3-eksbuild.x too for compatibility. Although I'm assuming it's for kubeadm, may not affect users on eks coredns addon.

**Comment by iul1an** [comment]
Hi @oliviassss 
Yes, you're right, I was actually referring to the Kubernetes node taints and node labels in this issue description, but I said tolerations :man_facepalming: 
Thank you!

**Comment by sjastis** [comment]
This is also fixed in coredns v1.10.1-eksbuild.5, v1.9.3-eksbuild.9 and v1.8.7-eksbuild.8.
